pipeline/dwi_masking.py

```diff
@@ -23,78 +24,110 @@
 
+# Use pure python implementation of protocol buffers
+# check if protobuf is later then 3.20.0 and if so, set the environment variable
+import pkg_resources
+
+if pkg_resources.get_distribution("protobuf").version >= "3.20.0":
+    os.environ['PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION'] = 'python'
+
 # Get the first available GPU
 try:
     import GPUtil
+
     DEVICE_ID_LIST = [g.id for g in GPUtil.getGPUs()]
     DEVICE_ID = DEVICE_ID_LIST[0]
-    
-    
-    CUDA_VISIBLE_DEVICES=os.getenv('CUDA_VISIBLE_DEVICES')
+
+    CUDA_VISIBLE_DEVICES = os.getenv('CUDA_VISIBLE_DEVICES')
     if CUDA_VISIBLE_DEVICES:
         # prioritize external definition
         if int(CUDA_VISIBLE_DEVICES) in DEVICE_ID_LIST:
             pass
         else:
             # define it internally
-            CUDA_VISIBLE_DEVICES=DEVICE_ID
+            CUDA_VISIBLE_DEVICES = DEVICE_ID
     else:
         # define it internally
-        CUDA_VISIBLE_DEVICES=DEVICE_ID
-    
+        CUDA_VISIBLE_DEVICES = DEVICE_ID
+
     os.environ["CUDA_VISIBLE_DEVICES"] = str(CUDA_VISIBLE_DEVICES)
     # setting of CUDA_VISIBLE_DEVICES also masks out all other GPUs
-    
-    print ("Use GPU device #", CUDA_VISIBLE_DEVICES)
 
+    print("Use GPU device #", CUDA_VISIBLE_DEVICES)
 
 except:
     os.environ["CUDA_VISIBLE_DEVICES"] = "-1"
     print("GPU not available...")
 
-
 import warnings
+
 with warnings.catch_warnings():
     warnings.filterwarnings("ignore", category=FutureWarning)
     import tensorflow as tf
 
+    # check version of tf and if 1.12 or less use tf.logging.set_verbosity(tf.logging.ERROR)
+    if int(tf.__version__.split('.')[0]) <= 1 and int(tf.__version__.split('.')[1]) <= 12:
+        tf.logging.set_verbosity(tf.logging.ERROR)
+        # Configure for dynamic GPU memory usage
+        config = tf.ConfigProto()
+        config.gpu_options.allow_growth = True
+        config.log_device_placement = False
+        sess = tf.Session(config=config)
+        try:
+            from keras import backend as K
+        except ImportError:
+            from tensorflow.keras import backend as K
+        K.set_session(sess)
+    else:
+        print("Num GPUs Available: ",
+              len(tf.config.experimental.list_physical_devices('GPU')))
+        tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)
+        gpus = tf.config.experimental.list_physical_devices('GPU')
+        if gpus:
+            try:
+                # Currently, memory growth needs to be the same across GPUs
+                for gpu in gpus:
+                    tf.config.experimental.set_memory_growth(gpu, True)
+                logical_gpus = tf.config.experimental.list_logical_devices(
+                    'GPU')
+                print(len(gpus), "Physical GPUs,", len(logical_gpus),
+                      "Logical GPUs")
+            except RuntimeError as e:
+                # Memory growth must be set before GPUs have been initialized
+                print(e)
```

Comment:
Do lines 79-108 solve the GPU memory issue? @RyanZurrin

Reply:
Yes, these are what keeps it from using the entire GPU Memory up. There are a couple of print statements that may not be needed anymore, I just realized. I was using them during debugging and for info while running.

Reply:
Can you omit them?

Reply:
yes. I will omit them now